Data/Unicode/Internal/NormalizeStream.hs

```diff
@@ -169,6 +178,52 @@ stream (Text arr off len) = Stream next off (betweenSize (len `shiftR` 1) len)
 {-# INLINE [0] stream #-}
 #endif
 
+-- | Perform Unicode normalization on @Text@ according to the specified
+-- normalization mode, using the quick check algorithm to speed up case when 
+-- the text is already normalized.
+normalizeQC
+    :: NormalizationMode
+    -> Text
+    -> Text
+normalizeQC mode t = case stream t of
+    str@(Stream next0 s0 _len) ->
+        let quickCheck s cc r = case next0 s of
+                Done       -> r
+                Skip    s' -> quickCheck s' cc r
+                Yield c s' ->
+                    -- [TODO] the ASCII check speeds up latin scripts,
+                    --        but it slows down the other scripts.
+                    if c <= '\x7f'
+                        -- For ASCII we know it’s always allowed and a starter
+                        then quickCheck s' 0 r
+                        -- Otherwise, lookup the combining class and QC property
+                        else let cc' = UC.combiningClass c in
+                            if cc > cc' && cc' /= 0
+                                then No
+                                else case check c of
+                                    No    -> No
+                                    Maybe -> quickCheck s' cc' Maybe
+                                    Yes   -> quickCheck s' cc' r
+                    -- let cc' = UC.combiningClass c
+                    -- in if cc > cc' && cc' /= 0
+                    --     then No
+                    --     else case check c of
+                    --         No    -> No
+                    --         Maybe -> quickCheck s' cc' Maybe
+                    --         Yes   -> quickCheck s' cc' r
+            check = case mode of
+                NFD  -> isNFD
+                NFKD -> isNFKD
+                NFC  -> isNFC
+                NFKC -> isNFKC
+        in case quickCheck s0 0 Yes of
+            Yes -> t
+            _   -> case mode of
```

Comment:
This won't give us any real benefit of quickcheck. After going through the entire string if we find that there is even one out-of-order case then we have to normalize the entire string.

Instead we need to keep doing quickcheck until we find an out-of-order case, then keep the buffer that we found is already normalized and go back up to the previous starter character and normalize the out-of-order case, then again start doing quickcheck and so on. It may be easier to tweak our regular normalization code to introduce quickcheck in that itself.

Also, we need special benchmarks for measuring the benefit in the best case, we need to run normalization on normalized or almost normalized inputs. We can save the normalized versions of our test inputs and try normalization on that. With quick check it should be significantly faster compared to normalizing it entirely.